### PR TITLE
feature: iio

### PR DESCRIPTION
**Pull request by adamhorden**
# **feature:** iio

This pull request is related to GNU Radio Enhancement Proposals 0017. In short it adds the gr-iio out of tree module in tree.

Please see GNU Radio Enhancement Proposals 0017:

https://github.com/gnuradio/greps/blob/master/grep-0017-iio.md

This is a rebase of the source against the original source from Travis Collins.

Please see:

https://github.com/analogdevicesinc/gnuradio/pull/4
https://github.com/analogdevicesinc/gnuradio/pull/5
https://github.com/analogdevicesinc/gnuradio/tree/merge-griio-upstream
https://github.com/gnuradio/gnuradio/pull/2840

**We have added fixes to allow GNU Radio version >= 3.9 compatibility:**

**Updates for GNU Radio version >= 3.9 compatibility.**

- Add pybind11 bindings.
- Use std::shared_ptr with GNU Radio version 3.9.
- Update CMakeLists to remove Volk.

:boom:

**Use std::shared_ptr with GNU Radio 3.9.**

In the ~~upcoming~~ now released, GNU Radio version 3.9 release, GNU Radio is switching from `boost::shared_pt`r to `std::shared_ptr`. 

:boom:

**Update CMakeLists to remove Volk.**

Remove the flag `ENABLE_VOLK` from CMakeLists.txt.

:boom:

**Testing**

I have tested these changes by building GNU Radio version 3.9 (master) from source and applying these changes. The build works as expected. I can supply a Docker container that allows this testing to happen in an automated manor.

```
#33 27.63 -- Found libad9361 library: /usr/lib/libad9361.so
#33 27.63 -- Found libiio library: /usr/lib/libiio.so
#33 27.63 --
#33 27.63 -- Configuring gr-iio support...
#33 27.63 --   Dependency Boost_FOUND = TRUE
#33 27.63 --   Dependency ENABLE_GNURADIO_RUNTIME = ON
#33 27.63 --   Dependency ENABLE_GR_BLOCKS = ON
#33 27.63 --   Dependency LIBIIO_FOUND = TRUE
#33 27.63 --   Dependency LIBAD9361_FOUND = TRUE
#33 27.63 --   Enabling gr-iio support.
```

It was requested to start a fresh PR with all changes.

Co-authored-by: Adam Horden <adam.horden@horden.engineering>
Co-authored-by: Edward Kigwana <ekigwana@scires.com>
Signed-off-by: Travis Collins <travis.collins@analog.com>

### CODE REVIEW

File: cmake/Modules/FindAD9361.cmake [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Copyright 2021 Free Software Foundation, Inc.
```
Comment (mbr0wn) [inline cmake/Modules/FindAD9361.cmake:1]: @adamhorden  if you wrote those, please change them to 
```suggestion
# Copyright 2021 Adam Horden
```

If you took them over this way, then just leave them and ignore this comment.

File: gr-iio/include/gnuradio/iio/api.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * GNU Radio is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GNU Radio; see the file COPYING.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
```
Comment (mbr0wn) [inline gr-iio/include/gnuradio/iio/api.h:20]: This should now be SPDX

Comment (mbr0wn) [reply gr-iio/include/gnuradio/iio/api.h:20]: (Same elsewhere, if applicable)

File: gr-iio/lib/device_source_impl.cc [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+}
+
+void device_source_impl::set_params(const std::vector<std::string>& params)
+{
+    set_params(this->phy, params);
+}
+
+void device_source_impl::set_buffer_size(unsigned int _buffer_size)
+{
+    boost::unique_lock<boost::mutex> lock(iio_mutex);
```
Comment (mbr0wn) [inline gr-iio/lib/device_source_impl.cc:100]: This is not a showstopper for merging IMO, but there's a lot of unnecessary boost usage here.

**Review by mbr0wn — APPROVED** [review]
I haven't even seen all the files yet -- but it's a new component with no cross-dependencies, merging this won't break anyone's code.

**Comment by willcode** [comment]
"Python bindings for attr_sink.h are out of sync". Others too?

File: cmake/Modules/FindAD9361.cmake (hunk 1 above)
Comment (adamhorden) [reply cmake/Modules/FindAD9361.cmake]: Thanks for spotting this 🤓 . 
I wrote these CMake build files to ensure a required dependancy on libiio and optional dependencies on libad9361.
Updated in a separate commit that can be rebased 😄 .

File: gr-iio/include/gnuradio/iio/api.h (hunk 2 above)
Comment (adamhorden) [reply gr-iio/include/gnuradio/iio/api.h]: Thanks for spotting this 🤓 . 
Updated to SPDX license headers.
Updated in a separate commit that can be rebased 😄 .

**Comment by adamhorden** [comment]
CI will need some updates. We will need some extra dependancies for gr_iio.

- libiio: https://github.com/analogdevicesinc/libiio
- libiio-ad9361: https://github.com/analogdevicesinc/libad9361-iio

I will take a look into this 🤓 

```
--   Dependency Boost_FOUND = TRUE
--   Dependency ENABLE_GNURADIO_RUNTIME = ON
--   Dependency ENABLE_GR_BLOCKS = ON
--   Dependency LIBIIO_FOUND = FALSE
--   Disabling gr-iio support.
--   Override with -DENABLE_GR_IIO=ON/OFF
```

**Comment by adamhorden** [comment]
> "Python bindings for attr_sink.h are out of sync". Others too?

This has now been fixed 🤓

**Comment by adamhorden** [comment]
While testing a bug :bug: was found when using GNU Radio Companion:

```
Unable to coerce endpoints: 'gnuradio.iio.iio_python.pluto_source' object has no attribute 'to_basic_block'
``` 

The issue was related to the pybind11 bindings. This has been fixed 🤓 . 
In case there is any one reviewing I have committed the fix as a separate commit that can be rebased later.

**Adam Horden**

File: gr-iio/lib/attr_source_impl.cc [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+                           gr_vector_const_void_star& input_items,
+                           gr_vector_void_star& output_items)
+{
+    int sample;
+    void* out;
+
+    out = output_items[0];
+
+    for (sample = 0; sample < samples_per_update; sample++) {
+        boost::this_thread::sleep_for(boost::chrono::milliseconds(update_interval_ms));
```
Comment (jacobagilbert) [inline gr-iio/lib/attr_source_impl.cc:256]: GR has moved to std::chrono, consider switching also to remove the dependency on boost::chrono. Suggest switching the includes to `<chrono>` and `<thread>` and using:

```c++
std::this_thread::sleep_for(std::chrono::milliseconds(update_interval_ms));
```

File: gr-iio/lib/attr_updater_impl.cc [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    d_thread->join();
+
+    return block::stop();
+}
+
+
+void attr_updater_impl::run()
+{
+    while (!d_finished) {
+        boost::this_thread::sleep(
```
Comment (jacobagilbert) [inline gr-iio/lib/attr_updater_impl.cc:106]: can use `std::this_thread::sleep`

File: gr-iio/python/iio/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+########################################################################
+# Handle the unit tests
+########################################################################
+if(ENABLE_TESTING)
+	set(GR_TEST_TARGET_DEPS "")
+	set(GR_TEST_LIBRARY_DIRS "")
+	set(GR_TEST_PYTHON_DIRS
+	  ${CMAKE_BINARY_DIR}/gnuradio-runtime/python
+ 	  ${CMAKE_BINARY_DIR}/gnuradio-runtime/swig
```
Comment (jacobagilbert) [inline gr-iio/python/iio/CMakeLists.txt:30]: No longer needed.

**Comment by jacobagilbert** [comment]
I am working through this, but with a clean master build on ubuntu 20.04 I get the following runtime error:

```python
>>> from gnuradio import iio
Traceback (most recent call last):
  File "/opt/gnuradio3-10/lib/python3/dist-packages/gnuradio/iio/__init__.py", line 31, in <module>
    from .iio_python import *
ImportError: /opt/gnuradio3-10/lib/python3/dist-packages/gnuradio/iio/iio_python.cpython-38-x86_64-linux-gnu.so: undefined symbol: _ZN2gr3iio10pluto_sink4makeERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEymmmbdPKcSB_ff
```

with
```
jacob@ubuntu:/opt/gnuradio/src/gnuradio/build$ c++filt _ZN2gr3iio10pluto_sink4makeERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEymmmbdPKcSB_ff
gr::iio::pluto_sink::make(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long long, unsigned long, unsigned long, unsigned long, bool, double, char const*, char const*, float, float)
```

**Comment by adamhorden** [comment]
Please see:

https://github.com/gnuradio/gnuradio/pull/4277

This branch has now been moved to the GNURadio organisation.

**Adam Horden**

**Pull request closed without merging**